src/components/SVGViewer/pinchzoom.ts

```diff
@@ -0,0 +1,853 @@
+  onDragUpdate: null,
+  onDoubleTap: null,
+};
+
+export default class PinchZoom {
+  // making everything public as it was all public originally and used by SVG viewer
+  public el: HTMLElement;
+  public container = document.createElement('div');
+  public zoomFactor = 1;
+  public lastScale = 1;
+  public offset = {
+    x: 0,
+    y: 0,
+  };
+  public initialOffset = {
+    x: 0,
+    y: 0,
+  };
+  public options: Options;
+  public isDoubleTap = false;
+  public enabled = false;
+
+  private lastDragPosition: Coordinates | null = null;
+  private lastZoomCenter: Coordinates | null = null;
+  private hasInteraction = false;
+  private nthZoom = 0;
+  private updatePlanned = false;
+  private inAnimation = false;
+  private _isOffsetsSet = false;
+
+  constructor(el: HTMLElement, options: Partial<Options>) {
+    this.el = el;
+
+    this.options = Object.assign({}, defaultOptions, options);
+    this.setupMarkup();
+    this.bindEvents();
+    this.update();
+
+    // The image may already be loaded when PinchZoom is initialized,
+    // and then the load event (which trigger update) will never fire.
+    if (this.isImageLoaded(this.el)) {
+      this.updateAspectRatio();
+      this.setupOffsets();
+    }
+
+    this.enable();
+  }
+
+  public isCloseTo(value: number, expected: number) {
+    return value > expected - 0.01 && value < expected + 0.01;
+  }
+
+  /**
+   * Event handler for 'dragstart'
+   * @param event
+   */
+  public handleDragStart(event: TouchEvent) {
+    if (typeof this.options.onDragStart == 'function') {
+      this.options.onDragStart(this, event);
+    }
+    this.stopAnimation();
+    this.lastDragPosition = null;
+    this.hasInteraction = true;
+    this.handleDrag(event);
+  }
+
+  /**
+   * Event handler for 'drag'
+   * @param event
+   */
+  public handleDrag(event: TouchEvent) {
+    const touch = this.getTouches(event)[0];
+    this.drag(touch, this.lastDragPosition);
+    this.offset = this.sanitizeOffset(this.offset);
+    this.lastDragPosition = touch;
+  }
+
+  public handleDragEnd(event: TouchEvent) {
+    if (typeof this.options.onDragEnd == 'function') {
+      this.options.onDragEnd(this, event);
+    }
+    this.end();
+  }
+
+  /**
+   * Event handler for 'zoomstart'
+   * @param event
+   */
+  public handleZoomStart(event: TouchEvent) {
+    if (typeof this.options.onZoomStart == 'function') {
+      this.options.onZoomStart(this, event);
+    }
+    this.stopAnimation();
+    this.lastScale = 1;
+    this.nthZoom = 0;
+    this.lastZoomCenter = null;
+    this.hasInteraction = true;
+  }
+
+  /**
+   * Event handler for 'zoom'
+   * @param event
+   * @param newScale
+   */
+  public handleZoom(event: TouchEvent, newScale: number) {
+    // a relative scale factor is used
+    const touchCenter = this.getTouchCenter(this.getTouches(event)),
+      scale = newScale / this.lastScale;
+    this.lastScale = newScale;
+
+    // the first touch events are thrown away since they are not precise
```

Comment:
that's interesting -- any idea why they're not precise? Assuming this is the code you copied over so maybe there's no answer. just curious.

Reply:
ah, no idea. I actually wanted to play a bit with it as well but will just leave it for now, as I spent a lot of time already.